File: leetcode/1957-delete-characters-to-make-fancy-string/solution.cpp
/**
 * Author: 1chooo<hugo970217@gmail.com>
 * Problem: https://leetcode.com/problems/delete-characters-to-make-fancy-string
 * Runtime: 15ms (93.51%)
 */

const static auto _ = []() {
    cin.tie(nullptr)->sync_with_stdio(false);
    return nullptr;
}();

class Solution {
  public:
    string makeFancyString(string s) {
        int count = 1;
        int len = s.size();
        string ans = "";
        ans.push_back(s[0]);

        for (int i = 1; i < len; i++) {
            if (s[i] != ans.back()) {
                count = 1;
                ans.push_back(s[i]);
            } else if (++count < 3) {
                ans.push_back(s[i]);
            }
        }

        return ans;
    }
};


/**
 * Author: 1chooo<hugo970217@gmail.com>
 * Problem: https://leetcode.com/problems/delete-characters-to-make-fancy-string
 * Runtime: 36ms (28.57%)
 */

const static auto _ = []() {
    cin.tie(nullptr)->sync_with_stdio(false);
    return nullptr;
}();

class Solution {
public:
    string makeFancyString(string s) {
        int count = 1;
        int len = s.size();
        string ans = "";
        ans.push_back(s[0]);

        for (int i = 1; i < len; i++) {
            if (s[i] == ans.back()) {
                count++;
                if (count < 3) {
                    ans.push_back(s[i]);
                }
            } else {
                count = 1;
                ans.push_back(s[i]);
            }
        }

        return ans;
    }
};

File: leetcode/1957-delete-characters-to-make-fancy-string/solution.py
"""
* Author: 1chooo<hugo970217@gmail.com>
* Problem: https://leetcode.com/problems/delete-characters-to-make-fancy-string
* Runtime: 289ms (70.27%)
"""

class Solution:
    def makeFancyString(self, s: str) -> str:
        if not s:
            return ""

        res = [s[0]]
        count = 1

        for i in range(1, len(s)):
            if s[i] != s[i - 1]:
                count = 1
                res.append(s[i])
            else:
                count += 1
                if count < 3:
                    res.append(s[i])

        return ''.join(res)
